### ISSUE TITLE
NetworkManager's nm-online kills nixos-rebuild

### ISSUE THREAD

**Issue by tbidne** [NixOS/nixpkgs#180175]
### Describe the bug

The systemd service `NetworkManager-wait-online.service` can prevent `nixos-rebuild` from succeeding:

```
warning: the following units failed: NetworkManager-wait-online.service

× NetworkManager-wait-online.service - Network Manager Wait Online
     Loaded: loaded (/etc/systemd/system/NetworkManager-wait-online.service; enabled; vendor preset: enabled)
    Drop-In: /nix/store/k5yq51spcggip2h6aq1y0bydkpr4zahc-system-units/NetworkManager-wait-online.service.d
             └─overrides.conf
     Active: failed (Result: exit-code) since Tue 2022-07-05 10:18:52 NZST; 36ms ago
       Docs: man:nm-online(1)
    Process: 1258376 ExecStart=/nix/store/b4yhg54s70i0v0k1qnnv8vnja6018yrh-networkmanager-1.38.2/bin/nm-online -s -q (code=exited, status=1/FAILURE)
   Main PID: 1258376 (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
        CPU: 22ms

Jul 05 10:17:52 nixos systemd[1]: Starting Network Manager Wait Online...
Jul 05 10:18:52 nixos systemd[1]: NetworkManager-wait-online.service: Main process exited, code=exited, status=1/FAILURE
Jul 05 10:18:52 nixos systemd[1]: NetworkManager-wait-online.service: Failed with result 'exit-code'.
```

This service runs `nm-online -s -q`, and the `nm-online` man page says:

```
-s | --wait-for-startup
           Wait for NetworkManager startup to complete, rather than waiting for network connectivity specifically. Startup is
           considered complete once NetworkManager has activated (or attempted to activate) every auto-activate connection
           which is available given the current network state. This corresponds to the moment when NetworkManager logs "startup
           complete". This mode is generally only useful at boot time. After startup has completed, nm-online -s will just
           return immediately, regardless of the current network state.

           There are various ways to affect when startup complete is reached. For details see NetworkManager-wait-
           online.service(8).
```

> **This corresponds to the moment when NetworkManager logs "startup complete". This mode is generally only useful at boot time.**

I am not familiar with this tool, but my experience is that after my laptop has been up for some time (e.g. days), `nm-online` will often return an error code rather than correctly determine the network is up, thus killing any future `nixos-rebuild` commands.

### Steps To Reproduce

1. Get your machine in a state where `nm-online -s -q` does not return success (not sure how to do this on demand).
2. Witness `nixos-rebuild` failure.

### Expected behavior

`nixos-rebuild` should not fail due to an erroneous network check.

### Additional context

This is tricky as it is not a `nix` issue per se but rather an issue with a presumably flaky systemd service. It is easy enough to disable this service manually:

```nix
systemd.services.NetworkManager-wait-online.enable = false;
```

And perhaps this is the best solution. But a number of my coworkers all ran into this issue independently, so I thought it merited an issue for discoverability, if nothing else. My gut reaction is that a flaky check should probably not be required by default, but I don't know enough about this service's importance/fragility to say.

This issue was noticed only recently, both on `nixos-unstable` and `nixos-22.05`.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.47, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Thanks!

**Comment by lorenz** [comment]
This is related to udev not initializing devices. NetworkManager never completes startup because a WireGuard interface is never initialized by udev. A workaround is just putting the affected device into `networking.networkmanager.unmanaged`.

**Comment by DeskworkTrickster** [comment]
I'm wondering if there is any solution to this. Since this triggers for me always. So, I'm not even sure if my `nixos-rebuild switch` does actually complete.

As mentioned in the initial description of the issue:
> This corresponds to the moment when NetworkManager logs "startup complete". This mode is generally only useful at boot time.

This never seems to be the case, since the actual message is "Started Network Manager.".
Apart from that I'm wondering why it relies on a log message string, when the status of `NetworkManager.service` would be way less prone to errors.

**Comment by oati** [comment]
I think I'm running into this.

`systemd-networkd-wait-online.service`
```
Aug 22 08:38:49 erin-laptop systemd[1]: Starting Wait for Network to be Configured...
Aug 22 08:40:49 erin-laptop systemd-networkd-wait-online[26249]: Timeout occurred while waiting for network connectivity.
Aug 22 08:40:49 erin-laptop systemd[1]: systemd-networkd-wait-online.service: Main process exited, code=exited, status=1/FAILURE
Aug 22 08:40:49 erin-laptop systemd[1]: systemd-networkd-wait-online.service: Failed with result 'exit-code'.
Aug 22 08:40:49 erin-laptop systemd[1]: Failed to start Wait for Network to be Configured.
```

It only happens after I've connected my USB-C dock with an ethernet connection at least once after boot. (Note: I'm running tmpfs on root, so my system should "forget" everything about my dock on reboot.)

**Comment by NorfairKing** [comment]
@stale don't you dare!

**Comment by pjones** [comment]
@ncfavier Any chance you can help with this?

**Comment by ncfavier** [comment]
I don't use NetworkManager so I wouldn't know, but in the case of systemd-networkd there are relevant options under `systemd.network.wait-online`: [`anyInterface`](https://search.nixos.org/options?channel=unstable&show=systemd.network.wait-online.anyInterface&from=0&size=50&sort=alpha_asc&type=packages&query=anyInterface) and [`ignoredInterfaces`](https://search.nixos.org/options?channel=unstable&show=systemd.network.wait-online.ignoredInterfaces&from=0&size=50&sort=alpha_asc&type=packages&query=ignoredInterfaces). I recommend at least setting the former to `true` on laptops.

> So, I'm not even sure if my `nixos-rebuild switch` does actually complete.

Warning about failed units is pretty much [the last thing](https://github.com/NixOS/nixpkgs/blob/f77a8a1df0cb83542be93f7d7ae962ab5a3ee78b/nixos/modules/system/activation/switch-to-configuration.pl#L935-L940) that the activation script does, so it's probably fine (but the failure should be fixed, of course).

**Comment by lorenz** [comment]
BTW I've "fixed" this by setting
```nix
# udev 250 doesn't reliably reinitialize devices after restart
systemd.services.systemd-udevd.restartIfChanged = false;
```

But this is really an upstream systemd bug.

**Comment by blaggacao** [comment]
My Ubuntu has:

```
❯ systemctl cat NetworkManager-wait-online.service
# /lib/systemd/system/NetworkManager-wait-online.service
[Unit]
Description=Network Manager Wait Online
Documentation=man:nm-online(1)
Requires=NetworkManager.service
After=NetworkManager.service
Before=network-online.target

[Service]
# `nm-online -s` waits until the point when NetworkManager logs
# "startup complete". That is when startup actions are settled and
# devices and profiles reached a conclusive activated or deactivated
# state. It depends on which profiles are configured to autoconnect and
# also depends on profile settings like ipv4.may-fail/ipv6.may-fail,
# which affect when a profile is considered fully activated.
# Check NetworkManager logs to find out why wait-online takes a certain
# time.

Type=oneshot
ExecStart=/usr/bin/nm-online -s -q
RemainAfterExit=yes

# Set $NM_ONLINE_TIMEOUT variable for timeout in seconds.
# Edit with `systemctl edit NetworkManager-wait-online`.
#
# Note, this timeout should commonly not be reached. If your boot
# gets delayed too long, then the solution is usually not to decrease
# the timeout, but to fix your setup so that the connected state
# gets reached earlier.
Environment=NM_ONLINE_TIMEOUT=60

[Install]
```

My latest NixOS config has:

```
> nix-repl> c.config.systemd.services.NetworkManager-wait-online
{ after = [ ... ]; aliases = [ ... ]; before = [ ... ]; bindsTo = [ ... ]; confinement = { ... }; conflicts = [ ... ]; description = ""; documentation = [ ... ]; enable = false; environment = { ... }; jobScripts = [ ... ]; onFailure = [ ... ]; partOf = [ ... ]; path = [ ... ]; postStart = ""; postStop = ""; preStart = ""; preStop = ""; reload = ""; reloadIfChanged = false; reloadTriggers = [ ... ]; requiredBy = [ ... ]; requires = [ ... ]; requisite = [ ... ]; restartIfChanged = true; restartTriggers = [ ... ]; runner = error: attribute 'ExecStart' missing

       at /nix/store/6dgpkrc0gxlndr4j2524ihlsr8209ph7-source/nixos/modules/testing/service-runner.nix:65:9:

           64|       my $cmd = <<END_CMD;
           65|       ${service.serviceConfig.ExecStart}
             |         ^
           66|       END_CMD
«derivation
```

based on [this definition](https://github.com/NixOS/nixpkgs/blob/2aa5c8d3abd627a871619260bc5cd83defa6c042/nixos/modules/services/networking/networkmanager.nix#L478):
```console
    systemd.services.NetworkManager-wait-online = {
      wantedBy = [ "network-online.target" ];
    };
```

Where is this `ExecStart` coming from in your configurations?

**Comment by bjornfor** [comment]
@blaggacao: The reference to `nm-online` comes from upstream service unit `NetworkManager-wait-online.service`, not from nixpkgs itself.

**Comment by pinpox** [comment]
I've been tripping over this bug for quite some time now and it is annoying for users. As mentioned above, the error can be worked around with:

```nix
systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;
systemd.services.systemd-networkd-wait-online.enable = lib.mkForce false;
```
I was concernd if there might be other dependencies or services that require this to be enabled, so I grepped through nixpkgs for both. These are the mentions:

### NetworkManager-wait-online.service
- [modules/networkmanager](https://github.com/NixOS/nixpkgs/blob/d23ce12697472b962de77f05bf28ebb1139f7b6a/nixos/modules/services/networking/networkmanager.nix#L478-L480) The actual definition of this service
- [pkgs.hqplayerd](https://github.com/NixOS/nixpkgs/blob/5b9f09ee63715405f7addd10625d2a68f7f9e62a/pkgs/servers/hqplayerd/default.nix#L96) can probably be removed

### systemd-networkd-wait-online.service
- [modules/system/boot/networkd](https://github.com/NixOS/nixpkgs/blob/d23ce12697472b962de77f05bf28ebb1139f7b6a/nixos/modules/system/boot/networkd.nix#L1904) Service devinition
- Used in various nixos tests, which should not be relevant to normal operation of the system (?)
  - nixos/tests/systemd-networkd-dhcpserver-static-leases.nix
  - nixos/tests/kea.nix
  - nixos/tests/systemd-networkd.nix
  - nixos/tests/systemd-bpf.nix
  - nixos/tests/systemd-networkd-dhcpserver.nix

### TL;DR

Upon first look the usage of these two services seem minimal to me and they are causing more problems that doing good. Agreeing with @domenkozar's proposal, I'd vote to disable them per default. If this is agreed upon, I can submit a PR

**Comment by matthiasbeyer** [comment]
I've been running with that service disabled for 6 months now and have not experienced a single issue. Don't count my voice too heavily, though :wink: ! :+1:

**Comment by lorenz** [comment]
If we're going to work around this I'd still prefer `systemd.services.systemd-udevd.restartIfChanged = false;` as the other workaround just masks the issue while udev's still half-broken.

**Comment by pinpox** [comment]
> If we're going to work around this I'd still prefer `systemd.services.systemd-udevd.restartIfChanged = false;` as the other workaround just masks the issue while udev's still half-broken.

For some reason, that didn't work for me. On rebuild it said "not restarting service"

**Comment by supermarin** [comment]
Yep confirming what @pinpox  said: 
```
updating GRUB 2 menu...
NOT restarting the following changed units: systemd-udevd.service
```

Still seeing the issue on HEAD. Disabling wait-online like mentioned previously fixes nixos-rebuild.
```
systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;
systemd.services.systemd-networkd-wait-online.enable = lib.mkForce false;
```

**Comment by ju1m** [comment]
`NetworkManager-wait-online.service` is a dependency of `network-online.target`:
```console
$ systemctl list-dependencies --reverse NetworkManager-wait-online.service
NetworkManager-wait-online.service
● └─network-online.target
●   ├─dnscrypt-proxy2.service
●   ├─wireguard-wg-intra-peer-[...].service
●   └─multi-user.target
```

```console
$ systemctl cat NetworkManager-wait-online.service | grep network-online.target
Before=network-online.target
WantedBy=network-online.target
```

As others pointed out, the `-s` (aka. `--wait-for-startup`) flag is causing some deadlock:
```
# `nm-online -s` waits until the point when NetworkManager logs
# "startup complete". That is when startup actions are settled and
# devices and profiles reached a conclusive activated or deactivated
# state. It depends on which profiles are configured to autoconnect and
# also depends on profile settings like ipv4.may-fail/ipv6.may-fail,
# which affect when a profile is considered fully activated.
# Check NetworkManager logs to find out why wait-online takes a certain
# time.
```
I guess using `-s` is a more accurate notification than simply waiting for `NetworkManager.service` to be up, and I don't know why it sometimes fails to notice that the startup is complete, nor whether removing it has any significant drawbacks, but keeping it causes `switch-to-configuration test` to fail often enough to bother me, so until a more correct fix is available, I'm personally removing it:
```nix
{ pkgs, ... }: {
  systemd.services.NetworkManager-wait-online = {
    serviceConfig.ExecStart = [ "" "${pkgs.networkmanager}/bin/nm-online -q" ];
  };
}
```

**Comment by darkone23** [comment]
Also running into these deadlocks on recent versions of nixpkgs.

**Comment by neilmayhew** [comment]
This has been happening to me consistently for over a year. The cause seems to be due to what I've described in https://github.com/NixOS/nixpkgs/issues/182449#issuecomment-1585296876.

I used (a simplified version of) @ju1m's workaround, but using the workaround from https://github.com/NixOS/nixpkgs/issues/195777#issuecomment-1324378856 seems to be even better. However, I don't know how to include that in my system configuration rather than running it manually when needed.

**Comment by Shados** [comment]
> This is related to udev not initializing devices. NetworkManager never completes startup because a WireGuard interface is never initialized by udev. A workaround is just putting the affected device into `networking.networkmanager.unmanaged`.

My instance of this issue is indeed caused by a wireguard interface, but adding the interface to `networking.networkmanager.unmanaged` (via name or `type:wireguard`) did not change a thing. Having done some more digging online, I believe I know why: prior to udev initialization completing for a link, NM cannot be sure what the name or type of the link will be (as udev may make changes), and so it is *not possible* to filter a device prior to udev marking it as configured.

I'm still looking for a workaround.

**Comment by martin-braun** [comment]
I have the issue on the MacBook Pro 2019 16,1 with t2 patches after switching to iwd, because of network issues.

**Comment by freekvh** [comment]
> I have the issue on the MacBook Pro 2019 16,1 with t2 patches after switching to iwd, because of network issues.
> 
> `systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;`
> 
> worked for me.

I'm new to NixOS, I have this error as well:

```bash
jul 20 19:19:13 nixos systemd[1]: Starting Network Manager Wait Online...
jul 20 19:20:13 nixos systemd[1]: NetworkManager-wait-online.service: Main process exited, code=exited, status=1/FAILURE
jul 20 19:20:13 nixos systemd[1]: NetworkManager-wait-online.service: Failed with result 'exit-code'.
jul 20 19:20:13 nixos systemd[1]: Failed to start Network Manager Wait Online.
warning: error(s) occurred while switching to the new configuration
```
 I put this line in my configuration.nix (`systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;`), but then I get this error:

```bash
$ sudo nixos-rebuild switch --upgrade
unpacking channels...
error: undefined variable 'lib'

       at /etc/nixos/configuration.nix:16:56:

           15|   boot.loader.systemd-boot.enable = true;
           16|   systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;
             |                                                        ^
           17|   boot.loader.efi.canTouchEfiVariables = true;
(use '--show-trace' to show detailed location information)
building Nix...
```

And I can't update anymore, so I am stuck and a fairly recent clean NixOS install. Is there a way to force the update for now? What is the best structural solution?

Thank you!

**Comment by bjornfor** [comment]
> I put this line in my configuration.nix (systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;), but then I get this error:

Looks like `nixos-generate-config` adds this in the beginning of /etc/nixos/configuration.nix:

```
[...snip comment block...]

{ config, pkgs, ... }:
```

whereas most people eventually modify that to `{ config, pkgs, lib, ... }:`, providing configuration.nix with the missing `lib` argument.

I think nixos-generate-config should be updated to include `lib` by default. Although the default config doesn't need/use `lib`, it quickly becomes something new NixOS users want to use.

**Comment by freekvh** [comment]
> > I put this line in my configuration.nix (systemd.services.NetworkManager-wait-online.enable = lib.mkForce false;), but then I get this error:
> 
> Looks like `nixos-generate-config` adds this in the beginning of /etc/nixos/configuration.nix:
> 
> ```
> [...snip comment block...]
> 
> { config, pkgs, ... }:
> ```
> 
> whereas most people eventually modify that to `{ config, pkgs, lib, ... }:`, providing configuration.nix with the missing `lib` argument.
> 
> I think nixos-generate-config should be updated to include `lib` by default. Although the default config doesn't need/use `lib`, it quickly becomes something new NixOS users want to use.

Thanx for the quic reply.

I have added { config, pkgs, **lib**, ... }:

In the first line of my configuration.nix, but I get the same error...

**Comment by bjornfor** [comment]
> I have added { config, pkgs, **lib**, ... }:
> 
> In the first line of my configuration.nix, but I get the same error...

Then please open a new issue or post on https://discourse.nixos.org/ asking for help.

**Comment by freekvh** [comment]
Ok, this solved it for me: https://discourse.nixos.org/t/nixos-rebuild-switch-upgrade-networkmanager-wait-online-service-failure/30746/2

Just dit `sudo nixos-rebuild boot --upgrade` and rebooted. I left the lib in the `{ config, pkgs, lib, ... }:`.

Thanx for the support!

**Comment by bjornfor** [comment]
> I think nixos-generate-config should be updated to include `lib` by default.

https://github.com/NixOS/nixpkgs/pull/244653

**Comment by Atemu** [comment]
So, to summarise, the actual error seems to be that NM waits for startup of some interfaces it thinks it should manage but they're never communicated as up to NM because NM was never supposed manage that interface?

When I say `nmcli`, NM thinks it does not manage my tailscale0 interface:

```
tailscale0: unmanaged
        "tailscale0"
        tun, sw, mtu 1280
```

Why would it be waiting on it then? The other interfaces are up and I'm sending you this message via one of them, so it can't be waiting on those.

I'll try to explicitly add it to `networking.networkmanager.unmanaged` and report back after a few rebuilds in a few weeks.

**Comment by K900** [comment]
No, `nm-online` is just weirdly broken and locks up if you try to do it on an already running system. It's probably an easy code fix but I'm not familiar with the codebase at all.

**Comment by pinpox** [comment]
> No, `nm-online` is just weirdly broken and locks up if you try to do it on an already running system. It's probably an easy code fix but I'm not familiar with the codebase at all.

Is that something that has to be fixed upstream? If so, could you create an issue for it?

**Comment by K900** [comment]
There's already an issue (kind of): https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1220

**Comment by pinpox** [comment]
> There's already an issue (kind of): https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1220 https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1034

Reading through those. Are we using the mentioned `-s` flag? If so, maybe we can remove it? Running `nm-online -s` manually hangs for me, while `nm-online` (without `-s`) seems to work correctly.

**Comment by K900** [comment]
We are, and we have to, at least on initial boot.

**Comment by pinpox** [comment]
Alright, I'll stick with my workaround then for now. I would like to contribute a real fix for this or sort it out properly, but don't know how to continue.

**Comment by neilmayhew** [comment]
I had this problem for well over a year, but since I added this to my `configuration.nix` the problem has not reoccurred:

```nix
systemd.services.NetworkManager-wait-online = {
  serviceConfig = {
    ExecStart = [ "" "${pkgs.networkmanager}/bin/nm-online -q" ];
  };
};
```

Nor have I had any other network problems.

Also,

```shellsession
$ nm-online -s
Connecting...............   30s [started]
$ journalctl -b -u NetworkManager-dispatcher.service
Jul 10 16:24:41 melchizedek systemd[1]: Starting Network Manager Script Dispatcher Service...
Jul 10 16:24:41 melchizedek systemd[1]: Started Network Manager Script Dispatcher Service.
Jul 10 16:24:51 melchizedek systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.
Jul 11 17:43:59 melchizedek systemd[1]: Starting Network Manager Script Dispatcher Service...
Jul 11 17:43:59 melchizedek systemd[1]: Started Network Manager Script Dispatcher Service.
Jul 11 17:44:09 melchizedek systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.
Jul 14 18:41:15 melchizedek systemd[1]: Starting Network Manager Script Dispatcher Service...
Jul 14 18:41:15 melchizedek systemd[1]: Started Network Manager Script Dispatcher Service.
Jul 14 18:41:25 melchizedek systemd[1]: NetworkManager-dispatcher.service: Deactivated successfully.
...
$ last reboot -1
reboot   system boot  5.15.119         Mon Jul 10 16:24   still running
```

`NetworkManager-dispatcher.service` seems to run at various times that aren't related to reboots or `nixos-upgrade switch`

**Comment by pinpox** [comment]
@neilmayhew what does that exactly do? If I understand the man page correctly, `-q` only omits the output, without changing any functionality

**Comment by neilmayhew** [comment]
@pinpox The `-q` was already there, and I'm just removing the `-s`. Sorry I didn't make that clearer.

With the extra configuration:

```ini
$ systemctl cat NetworkManager-wait-online.service
# /etc/systemd/system/NetworkManager-wait-online.service
[Unit]
Description=Network Manager Wait Online
Documentation=man:nm-online(1)
Requires=NetworkManager.service
After=NetworkManager.service
Before=network-online.target

[Service]
# `nm-online -s` waits until the point when NetworkManager logs
# "startup complete". That is when startup actions are settled and
# devices and profiles reached a conclusive activated or deactivated
# state. It depends on which profiles are configured to autoconnect and
# also depends on profile settings like ipv4.may-fail/ipv6.may-fail,
# which affect when a profile is considered fully activated.
# Check NetworkManager logs to find out why wait-online takes a certain
# time.

Type=oneshot
ExecStart=/nix/store/r7fwxkxbgmwyrff5n03r9nchzskj59n9-networkmanager-1.40.16/bin/nm-online -s -q
RemainAfterExit=yes

# Set $NM_ONLINE_TIMEOUT variable for timeout in seconds.
# Edit with `systemctl edit NetworkManager-wait-online`.
#
# Note, this timeout should commonly not be reached. If your boot
# gets delayed too long, then the solution is usually not to decrease
# the timeout, but to fix your setup so that the connected state
# gets reached earlier.
Environment=NM_ONLINE_TIMEOUT=60

[Install]
WantedBy=network-online.target

# /nix/store/2ma3yvlqmj94dr6v24nkwmspgxfpfnmq-system-units/NetworkManager-wait-online.service.d/overrides.conf
[Unit]

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/g135laybfk1c7bm2zhrh7x6dv884qxal-glibc-locales-2.35-224/lib/locale/locale-archive"
Environment="PATH=/nix/store/l6jgwxkc3jhr029vfzfwzcy28iyckwsj-coreutils-9.1/bin:/nix/store/gn1s1s5z19cf0wiir2cd38jckcjc6kn6-findutils-4.9.0/bin:/nix/store/pvb117r7fhwb08717ks21a6y>
Environment="TZDIR=/nix/store/z0kg1c0f8fx6r4rgg5bdy01lb2b9izqg-tzdata-2023a/share/zoneinfo"

ExecStart=
ExecStart=/nix/store/r7fwxkxbgmwyrff5n03r9nchzskj59n9-networkmanager-1.40.16/bin/nm-online -q
```

**Comment by lilyball** [comment]
Adding `tailscale0` to `networking.networkmanager.unmanaged` on my system that's currently experiencing this issue did not resolve the problem.

**Comment by AkechiShiro** [comment]
Is there any way we can properly fix this issue ?

**Comment by Atemu** [comment]
Not on the NixOS side; we'd have to fix it upstream I imagine.

**Comment by charmoniumQ** [comment]
Is the problem that `nm-online` with `-s` on an already-up system hangs (what this thread seems to think)?

Is the problem that `NetworkManager-wait-online.service` _should_ not have `-s` (what [upstream](https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/issues/1220) seems to think)?

**Comment by Atemu** [comment]
I do not see any confusion about this from the upstream side; they did not claim that distros should not use `-s` at any point in the thread you linked. 

We want a flag that does what `-s` is currently documented to supposedly do:

```
 -s | --wait-for-startup
     Wait for NetworkManager startup to complete, rather than waiting for network connectivity specifically. Startup is considered complete once NetworkManager has activated (or
     attempted to activate) every auto-activate connection which is available given the current network state. This corresponds to the moment when NetworkManager logs "startup
     complete". This mode is generally only useful at boot time. After startup has completed, nm-online -s will just return immediately, regardless of the current network state.

     There are various ways to affect when startup complete is reached. For details see NetworkManager-wait-online.service(8).
```

The manual even explicitly clarifies behaviour after bootup is complete.

This is precisely what we want but `nm-online` simply does not work as documented. Instead of returning immediately, it blocks indefinitely under certain conditions.

**Comment by inmaldrerah** [comment]
Is this related to kernel/firmware upgrade? This problem just occurred on my machine after a `nixos-rebuild switch`, manually invoking `nm-online -s` also waited 30s and timed out, but after a reboot I can no longer reproduce this, even with another `nixos-rebuild switch`. I updated my kernel and firmware in the first `nixos-rebuild switch`, updating from linux 6.8 to 6.9.

**Comment by K900** [comment]
No, it's not.

**Comment by inmaldrerah** [comment]
I noted that `tailscaled.service` is restarted before `NetworkManager.service` and `NetworkManager-wait-online.service`.
```
restarting the following units: home-manager-inme.service, home-manager-root.service, nix-daemon.service, polkit.service, systemd-journald.service, tailscaled.service
starting the following units: NetworkManager-wait-online.service, NetworkManager.service, audit.service, bluetooth.service, kmod-static-nodes.service, logrotate-checkconf.service, mount-pstore.service, network-local-commands.service, network-setup.service, persist-\x27-nix-persist-etc-machine\x2did\x27.service, persist-\x27-nix-persist-home-inme-.bash_history\x27.service, persist-\x27-nix-persist-home-inme-.gitconfig\x27.service, persist-\x27-nix-persist-home-inme-.repo_.gitconfig.json\x27.service, power-profiles-daemon.service, resolvconf.service, rtkit-daemon.service, systemd-hostnamed.service, systemd-modules-load.service, systemd-oomd.socket, systemd-sysctl.service, systemd-timesyncd.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-update-done.service, systemd-vconsole-setup.service, waydroid-container.service, wpa_supplicant.service, xray.service
```
Also confirmed in journalctl:
<details>

```
May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Started Tailscale node agent.
░░ Subject: A start job for unit tailscaled.service has finished successfully
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit tailscaled.service has finished successfully.
░░ 
░░ The job identifier is 3587.
...
May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Starting Network Manager...
░░ Subject: A start job for unit NetworkManager.service has begun execution
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit NetworkManager.service has begun execution.
░░ 
░░ The job identifier is 3744.
...
May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Starting Network Manager Wait Online...
░░ Subject: A start job for unit NetworkManager-wait-online.service has begun execution
░░ Defined-By: systemd
░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
░░ 
░░ A start job for unit NetworkManager-wait-online.service has begun execution.
░░ 
░░ The job identifier is 3743.
```
</details>

And if I `systemctl stop tailscaled` before `nixos-rebuild switch`, `tailscaled.service` will be started after `NetworkManager.service` is started, and then `NetworkManager-wait-online.service` will exit gracefully.

**Comment by Atemu** [comment]
Big if true.

That feels like something that should be solvable using a few systemd Before/After directives in the right places.

Though tailscaled.service already contains

```
After=network-pre.target NetworkManager.service systemd-resolved.service
```

IIRC `After` should be applied in reverse when stopping/restarting.

**Comment by Vaisakhkm2625** [comment]
for me
```sh
sudo nixos-rebuild boot --flake .#nixos
```
instead of 

```sh
sudo nixos-rebuild switch --flake .#nixos
```
did the trick..

**Comment by AkechiShiro** [comment]
@atemu could you challenge upstream about the `-s` flag and its documented behavior ?

**Comment by jbury** [comment]
> Big if true.
> 
> That feels like something that should be solvable using a few systemd Before/After directives in the right places.
> 
> Though tailscaled.service already contains
> 
> ```
> After=network-pre.target NetworkManager.service systemd-resolved.service
> ```
> 
> IIRC `After` should be applied in reverse when stopping/restarting.

Just to add a +1 to the tailscale discussion - I did not have this issue at all prior to installing tailscale, and now have it any time a rebuild needs to restart NetworkManager.

**Comment by inmaldrerah** [comment]
> `After` should be applied in reverse when stopping/restarting.

It seems currently [the switching script (the Perl version)](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/activation/switch-to-configuration.pl#L938) doesn't have such mechanisms, neither does [the Rust version](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/sw/switch-to-configuration-ng/src/main.rs#L1763). For the Perl script, it just `systemctl restart`s and then `systemctl start`s. Meanwhile the Rust version seems to be [waiting](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/sw/switch-to-configuration-ng/src/main.rs#L1785) all `restart` threads to finish before starting the `start` threads.

A probable solution to this is to use combine these two commands into one `systemctl restart`, as stopped services will just get started on `systemctl restart`, and systemctl/systemd should be able (?) to resolve these Before/After relations if we invoke the command in one go. For the rust version, maybe removing that `block_on_jobs` and changing accordingly will be sufficient.

—— in case this is really the cause of the problem.

**Comment by howyay** [comment]
> I noted that `tailscaled.service` is restarted before `NetworkManager.service` and `NetworkManager-wait-online.service`.
> 
> ```
> restarting the following units: home-manager-inme.service, home-manager-root.service, nix-daemon.service, polkit.service, systemd-journald.service, tailscaled.service
> starting the following units: NetworkManager-wait-online.service, NetworkManager.service, audit.service, bluetooth.service, kmod-static-nodes.service, logrotate-checkconf.service, mount-pstore.service, network-local-commands.service, network-setup.service, persist-\x27-nix-persist-etc-machine\x2did\x27.service, persist-\x27-nix-persist-home-inme-.bash_history\x27.service, persist-\x27-nix-persist-home-inme-.gitconfig\x27.service, persist-\x27-nix-persist-home-inme-.repo_.gitconfig.json\x27.service, power-profiles-daemon.service, resolvconf.service, rtkit-daemon.service, systemd-hostnamed.service, systemd-modules-load.service, systemd-oomd.socket, systemd-sysctl.service, systemd-timesyncd.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-update-done.service, systemd-vconsole-setup.service, waydroid-container.service, wpa_supplicant.service
> ```
> 
> Also confirmed in journalctl:
> 
> ```
> May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Started Tailscale node agent.
> ░░ Subject: A start job for unit tailscaled.service has finished successfully
> ░░ Defined-By: systemd
> ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
> ░░ 
> ░░ A start job for unit tailscaled.service has finished successfully.
> ░░ 
> ░░ The job identifier is 3587.
> ...
> May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Starting Network Manager...
> ░░ Subject: A start job for unit NetworkManager.service has begun execution
> ░░ Defined-By: systemd
> ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
> ░░ 
> ░░ A start job for unit NetworkManager.service has begun execution.
> ░░ 
> ░░ The job identifier is 3744.
> ...
> May 28 07:26:58 thinkbook-16-plus-nixos systemd[1]: Starting Network Manager Wait Online...
> ░░ Subject: A start job for unit NetworkManager-wait-online.service has begun execution
> ░░ Defined-By: systemd
> ░░ Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
> ░░ 
> ░░ A start job for unit NetworkManager-wait-online.service has begun execution.
> ░░ 
> ░░ The job identifier is 3743.
> ```
> 
> And if I `systemctl stop tailscaled` before `nixos-rebuild switch`, `tailscaled.service` will be started after `NetworkManager.service` is started, and then `NetworkManager-wait-online.service` will exit gracefully.

Thank you for this observation - adding `systemd.services.tailscaled.after = ["NetworkManager-wait-online.service"]` seem to have solved the issue for me

**Comment by Atemu** [comment]
I'd invite everyone to try this out on their machines and report back.

If this works, that'd be a quick and easy one-line patch to the tailscale NixOS module; fixing this for everyone.

**Comment by supermarin** [comment]
Can confirm this fixes it. @Atemu +1 for including the fix in the tailscale module

**Comment by supermarin** [comment]
Opened a PR: https://github.com/NixOS/nixpkgs/pull/344678 
Testing and feedback appreciated

**Comment by AkechiShiro** [comment]
Is this only fixed when using tailscaled, for users not using tailscaled, they should apply the workaround mentioned ? Is this really a good solution for new users?

**Comment by supermarin** [comment]
@AkechiShiro have you observed this without using Tailscale? 

As far as I understand the issue is tied down to the combination of Tailscale firing up before NetworkManager and depending on it being online, ending up in a deadlock essentially

**Comment by pinpox** [comment]
The same issue happened to me with wireguard in the past. I'll try the suggested fix

**Comment by AkechiShiro** [comment]
@supermarin I have yes and I believe I'm not the only one

**Comment by supermarin** [comment]
Mind posting your configuration.nix in a gist?
We should be able to create a minimal flake and reproduce this with build-vm. Maybe cut out the VM's network access on purpose.

We should reopen this issue then if this happens outside of just Tailscale & NM

**Comment by AkechiShiro** [comment]
I do use Wireguard yes, so it could be I have the same issue as @pinpox

**Comment by Atemu** [comment]
Can anyone confirm whether they still experience this issue without using tailscaled?

**Comment by sgraf812** [comment]
As recently as two years ago, I had this issue without knowingly using tailscaled. (Posted to #59603.)
Since then, I'm using the workaround from the OP and have lived happily ever since.

The problem is perhaps that tailscaled is not *the only* service that activates `NetworkManager-wait-online.service`, so it doesn't make sense to fix this issue in tailscaled.

Perhaps this issue has since been fixed by an unrelated patch upstream (however that would happen), but I'm not optimistic. Let me try to get rid of the workaround and see if it is still an issue.

**Comment by AkechiShiro** [comment]
Thanks @sgraf812 I may try to reach out to systemd/udev maintainers about issue #182449 hopefully we can nail this issue down once and for all

**Comment by AkechiShiro** [comment]
We did receive an answer from Poettering but I don't know how to properly answer back, if anyone more experienced could pitch in : https://github.com/systemd/systemd/issues/34585

**Comment by Atemu** [comment]
Just ran into this again with tailscaled.

**Comment by supermarin** [comment]
@Atemu ran into it as well, but already had a running system with tailscale up. Manually shut down tailscale, and rebuild succeeded (and tailscale was back up). Tested with several rebuilds and restarts, works ok so far

**Comment by Atemu** [comment]
An idea I had on this is that we could perhaps hack around this using a systemd unit that `Conflicts` the units which appear to block nm-online.target such as tailscaled. It'd be `Require`d by the nm-online target and we'd then have another target which wants all those units and is wanted by the network-online target but `After` the nm-online target.

**Comment by inmaldrerah** [comment]
Maybe [this](https://github.com/NixOS/nixpkgs/pull/337667/commits/c52df39ef87b337afaa040a128df9014bb96cf9c) would help? (I closed because earlier I thought it was obsolete)

And I've made a small flake version for testing purpose: https://github.com/inmaldrerah/nixos-extensions

At least last time I used this, I didn't have to `systemd stop tailscaled` beforehand when tailscale was updated.

**Comment by supermarin** [comment]
I can try testing with it.
Think the repro steps are:
1. Have a system with Tailscale and NM-wait-online 
2. Disable nm-wait-online with the hack above
3. Put nm-wait-online back in

On the next rebuild it should get stuck. I'll try this later today and report back if it reproduces consistently. Note: the Tailscale fix was backported to 24.05 so need to pin nixpkgs to a commit prior to that

**Comment by Atemu** [comment]
Note that this issue occurs even with the supposed fix.

**Comment by JonnieCache** [comment]
Also still experiencing this with tailscaled on nixpkgs unstable `bc947f541ae55e999ffdb4013441347d83b00feb` (I think that's the right sha.)

If I stop the tailscaled unit I can then rebuild OK, and the service comes back up again.

**Comment by Munksgaard** [comment]
I'm still running into this problem. However, I'm not using NetworkManager, but systemd-networkd. Adding the following line to my config seemed to fix things:

```nix
systemd.services.tailscaled.after = ["systemd-networkd-wait-online.service"];
```

Perhaps we should also add that line to the tailscale service definition?

**Comment by inmaldrerah** [comment]
> Yeah just ran into it as well :( @inmaldrerah mind posting what's the best way to test your flake? I looked into it quickly but not 100% clear on how to plug in switch-to-configuration-ng.

Well, I haven't been using it for a while as I am no longer experiencing this problem on the unstable branch until `a73246e2eef4c6ed172979932bc80e1404ba2d56`. Which commit are you on now?

That flake can be used by making an overlay over nixpkgs, disable `system.switch.enable` and enable `system.switch.enableNg` in the `configuration.nix`.

**Comment by afifurrohman-id** [comment]
Are the only current workaround is:
```nix
# /etc/nixos/configuration.nix
systemd.network.wait-online.enable = false;
```

**Comment by maikelthedev** [comment]
> Are the only current workaround is?:
> 
> # /etc/nixos/configuration.nix
> systemd.network.wait-online.enable = false;

It's actually

>systemd.services.NetworkManager-wait-online.enable 

anythign other than that would fail.

**Comment by afifurrohman-id** [comment]
> > Are the only current workaround is?:
> > # /etc/nixos/configuration.nix
> > systemd.network.wait-online.enable = false;
> 
> It's actually
> 
> > systemd.services.NetworkManager-wait-online.enable
> 
> anythign other than that would fail.

I use stable version `24.11` then why it's works?

https://search.nixos.org/options?channel=24.11&show=systemd.network.wait-online.enable&from=0&size=50&sort=relevance&type=packages&query=systemd.network.wait-online.enable

**Comment by ndarilek** [comment]
Hitting the same issue here. Adding `systemd.services.tailscaled.after = [ "systemd-networkd-wait-online.service" ];` fixed it for me too.

**Comment by esibun** [comment]
`systemd.services.tailscaled.after = ["NetworkManager-wait-online.service"]` stopped working for me today; just had a switch fail on me despite verifying that NetworkManager-wait-online.service was listed in the systemd override file.  The earlier workaround removing `-s` from the ExecStart continues to work fine.

I'm highly suspicious that this has to do with NetworkManager being updated; I'm not sure what dependency caused a rebuild in this particular case, but the NetworkManager hash did change for me between generations.  That might be the trigger to actually reproducing this.

**Comment by ahydronous** [comment]
I am also hitting this issue on `25.05` despite using
```nix
  systemd.services.NetworkManager-wait-online.enable = false;
  systemd.services.tailscaled.after = [ "systemd-networkd-wait-online.service" ];
```

**Comment by codethief** [comment]
I'm not sure this is the right ticket to post this in but my issue at least sounds very much related: I just ran into `NetworkManager-wait-online.service` blocking my graphical session ([uwsm](https://github.com/Vladimir-csp/uwsm)/Sway) for 60s upon boot (i.e. for the duration of the timeout) because the network couldn't be configured. The issue seems to be that I set `virtualisation.docker.enable = true;` (using the default `enableOnBoot = true;`) which results in the following effective chain of dependencies:

```
uwsm
▼ waits for (via polling)
graphical.target
▼ (Requires & After)
multi-user.target
▼ (Wants & After)
docker.service
▼ (Wants & After)
network-online.target
▼ (Wants & After)
NetworkManager-wait-online.service
```

So if `NetworkManager-wait-online.service` takes long (e.g. the full 60s till it times out), `graphical.target` will, too.
